### PR TITLE
feat: add Delta Lake support

### PR DESCRIPTION
**Pull request by roeap**
## Summary & Motivation

This PR is still work in progress, but since I expect that if this is a wanted feature it may have to go through a couple of review cycles, I thought it may already be worthwhile to open this for review.

This PR adds a new integration library `dagster-deltalake` to support reading from  and writing to tables stored in the [Delta](https://delta.io/) format. Alog with support for some data frame libraries via `dagster-deltalake-pandas` and `dagster-deltalake-polars`.

While in principle, one can work with Delta tables already using the PySpark integration, I hope that this is still a valuable contribution. In my personal expereince its a siginificant quality of life improvement not having to deal with spark in various environments.

In large parts I tried to mimic the behavior of the DuckDB integrations and rely on the `DbIOManager`. Since deltalake internally relies heavily on Arrow, I tried to centralize type handler logic via the `DeltalakeBaseArrowTypeHandler`. Other type handlers just implement conversions from and to arrow, which should be zero-copy for pyarrow and polars, and for pandas, once 2.0 is supported.

Since there is a large overlap in the supported features between the DuckDB integration and Delta Lake, I more or less copied the documentation for DuckDB and adjsuted it to reflect the deltalake apis.

There are a few more things that need documenting beyond the DuckDB features:

- [ ] connecting with cloud storages
- [ ] triggering jobs based on version updates
- [ ] Running SQL queries with DuckDB + Delta Lake

## How I Tested These Changes

- [x] port `dasgster-duckdb-*` test suite for type handlers
- [ ] additional tests specific for deltalake type handlers

### CODE REVIEW

**Comment by sryza** [comment]
Hey @roeap - it looks like we missed this!

@PedramNavid is currently working on some guidelines and infrastructure that will enable us to include externally-maintained integrations in our monorepo. After he completes those, we'd love to include this if you're up for maintaining it.

**Comment by roeap** [comment]
Great to hear @sryza! 

Obviously my hope would be that the community embraces this and there is some help from that end, but keeping up with Dagster updates once core functionality is stabilised is something I'd be happy to do.

As this would be externally maintained, one more question. Right now I started adding docs and examples analogous to duck-db, which would land fairly prominently (agian same as duck-db :)) in the docs. Is that something you'd support, or should they be less prominent?

Last but not least, is there a rough estimate when this tooling would land? Just asking so I can properly prioritise my time :)

**Comment by sryza** [comment]
> Right now I started adding docs and examples analogous to duck-db, which would land fairly prominently (agian same as duck-db :)) in the docs.

If you're referring to the docs content that's in this PR, I think this would definitely be helpful to include.

> Last but not least, is there a rough estimate when this tooling would land?

@PedramNavid and @jmsanders - do you have a rough estimate for when the external integrations guide and buildkite changes will land?

**Comment by jmsanders** [comment]
Public unit test builds are almost ready - just waiting on https://github.com/dagster-io/dagster/pull/15762 to merge plus a little bit more acceptance testing on my end. We might need to hop through a few hoops to get this existing PR to trigger a build (pushing a new commit potentially? or closing and re-opening?) but we'll cross that bridge when we get there.

**Comment by roeap** [comment]
@PedramNavid - picking this up again :).

Still need to do some updating and polishing and of yourde getting CI to green. Just merged your branch with the fixes, and latest master. 

I can follow the link to buildkite, but the run seems to be blocked for some reason. May this be b/c of the failing previous step? There I have a feeling it may somehow be credential releated, at least I cannot see a specific failure.

Any help is appreciated :).

**Comment by sryza** [comment]
Hey @roeap - at the moment, buildkite requires a Dagster Labs engineer to unblock. I just unblocked it for you.

**Comment by roeap** [comment]
Hey @sryza , @PedramNavid - sorry I have been slow to respond.

Reserved some time this we and next week so we can hopefully push this over the finish line. Rebuild pyright evs and linters etc. Locally things look fine so far.

When it comes to the APIs, I have been following the duckdb implementaion. There are a few things in the implementation I'd love some feedback on.

* The configuration is quite complex, or at least seems like it since almost all options are exposed. Should this be simplified?
* I included arrow type support by default in the pandas handlers, since its a required dependency anyhow. should such a default be included?
* There are a few things I am still pondering - i.e. deltalake has some higher level APis to define different merge strategies when updating a partition. Is configuring these on the io manager a useful APi in your opinion? Probably also via the asset metadata.

Thanks for your help.

File: python_modules/libraries/dagster-deltalake-polars/MANIFEST.in [hunk 1]
```diff
@@ -0,0 +1,2 @@
+include LICENSE
+include dagster_deltalake_pandas/py.typed
```
Comment (PedramNavid) [inline python_modules/libraries/dagster-deltalake-polars/MANIFEST.in:2]: polars

File: python_modules/libraries/dagster-deltalake-polars/setup.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+from pathlib import Path
+from typing import Dict
+
+from setuptools import find_packages, setup
+
+
+def get_version() -> str:
+    version: Dict[str, str] = {}
+    with open(Path(__file__).parent / "dagster_deltalake_pandas/version.py", encoding="utf8") as fp:
```
Comment (PedramNavid) [inline python_modules/libraries/dagster-deltalake-polars/setup.py:9]: polars

File: python_modules/libraries/dagster-deltalake/dagster_deltalake/io_manager.py [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+
+class _DeltaTableIOManagerResourceConfig(TypedDict):
+    root_uri: str
+    storage_options: _StorageOptionsConfig
+    client_options: NotRequired[Dict[str, str]]
+    table_config: NotRequired[Dict[str, str]]
+
+
+class DeltaLakeIOManager(ConfigurableIOManagerFactory):
+    """Base class for an IO manager definition that reads inputs from and writes outputs to DuckDB.
```
Comment (PedramNavid) [inline python_modules/libraries/dagster-deltalake/dagster_deltalake/io_manager.py:57]: DuckDB

**Review by PedramNavid — APPROVED** [review]

**Pull request merged**